### PR TITLE
Fix test-runner on FreeBSD

### PR DESCRIPTION
**Pull request by rincebrain**
### Motivation and Context
CLOCK_MONOTONIC_RAW is only a thing on Linux and macOS. I'm not
actually sure why the previous hardcoding of a constant didn't
error out, but when we removed it, it sure does now.

~(Separately, should figure out why it's returning success when erroring out in under a second in the CI, but wanted to get this in before possibly going down a rabbit hole like that...)~ Ah, zfs-tests.sh doesn't actually care if the test-runner dies in a fire, it just cares what the results collector returns...let's fix that.

### Description
Fallback to CLOCK_MONOTONIC if CLOCK_MONOTONIC_RAW doesn't exist.

Return the error code if test-runner.py errored, even if zts-results didn't.

### How Has This Been Tested?
Didn't run on my 13-RELEASE VM before, runs now.

### Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Performance enhancement (non-breaking change which improves efficiency)
- [ ] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Library ABI change (libzfs, libzfs\_core, libnvpair, libuutil and libzfsbootenv)
- [ ] Documentation (a change to man pages or other documentation)

### Checklist:

- [x] My code follows the OpenZFS [code style requirements](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md#coding-conventions).
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**contributing** document](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md).
- [ ] I have added [tests](https://github.com/openzfs/zfs/tree/master/tests) to cover my changes.
- [x] I have run the ZFS Test Suite with this change applied.
- [x] All commit messages are properly formatted and contain [`Signed-off-by`](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md#signed-off-by).

### CODE REVIEW

**Comment by amotin** [comment]
I am just curios, are effects of ntpd dramatic enough to need CLOCK_MONOTONIC_RAW here at all?

File: scripts/zfs-tests.sh [hunk 1]
```diff
@@ -695,17 +695,21 @@ fi
...
-${TEST_RUNNER} ${QUIET:+-q} \
+(${TEST_RUNNER} ${QUIET:+-q} \
     -c "${RUNFILES}" \
     -T "${TAGS}" \
     -i "${STF_SUITE}" \
     -I "${ITERATIONS}" \
-    2>&1 | tee "$RESULTS_FILE"
+    2>&1 ; echo $? > "$EVILRES"; )| tee "$RESULTS_FILE"
+
+RUNRESULT=$(cat "$EVILRES")
```
Comment (nabijaczleweli) [inline scripts/zfs-tests.sh:711]:
```sh
{ ${T_R}
    -...
2>&1; RUNRESULT=$?; } | tee "$R_F"
```
and you can drop the temp file

**Comment by rincebrain** [comment]
> I am just curios, are effects of ntpd dramatic enough to need CLOCK_MONOTONIC_RAW here at all?

I'd probably drop it unless someone chimes in with a justification, personally, I just didn't have any desire to start a debate over it versus fixing the immediate problem here.

File: scripts/zfs-tests.sh (hunk 1 above)
Comment (rincebrain) [reply scripts/zfs-tests.sh:711]: Tragically...
zfs-tests.sh:708:12: note: Modification of RUNRESULT is local (to subshell caused by pipeline). [SC2030]
zfs-tests.sh:753:7: note: RUNRESULT was modified in a subshell. That change might be lost. [SC2031]

So should I silence it, do you think?

**Comment by szubersk** [comment]
Thank you, @rincebrain , for tagging me. I just managed to hack my way to the issue reproduction.

```
+ /usr/home/dev/zfs/scripts/zfs-tests.sh -vx -d /mnt/ -I 1 -T functional
+ RESULT=0
+ set +x

[dev@vm ~/test]$ cat test.log 

--- Cleanup ---
Removing pool(s):      
Removing loopback(s):  
Removing files(s):     

--- Configuration ---
Runfiles:        /home/dev/zfs/tests/runfiles/common.run,/home/dev/zfs/tests/runfiles/freebsd.run
STF_TOOLS:       /home/dev/zfs/tests/test-runner
STF_SUITE:       /home/dev/zfs/tests/zfs-tests
STF_PATH:        /home/dev/zfs/bin
FILEDIR:         /mnt/
FILES:           /mnt//file-vdev0 /mnt//file-vdev1 /mnt//file-vdev2
LOOPBACKS:       md0 md1 md2
DISKS:           md0 md1 md2
NUM_DISKS:       3
FILESIZE:        4G
ITERATIONS:      1
TAGS:            functional
STACK_TRACER:    no
Keep pool(s):    rpool
Missing util(s): mount.zfs zed zgenhostid devname2devid getversion mmap_libaio randfree_file user_ns_exec xattrtest

/home/dev/zfs/tests/test-runner/bin/test-runner.py  -c "/home/dev/zfs/tests/runfiles/common.run,/home/dev/zfs/tests/runfiles/freebsd.run" -T "functional" -i "/home/dev/zfs/tests/zfs-tests" -
I "1"
Traceback (most recent call last):
  File "/home/dev/zfs/tests/test-runner/bin/test-runner.py", line 35, in <module>
    from time import time, CLOCK_MONOTONIC_RAW
ImportError: cannot import name 'CLOCK_MONOTONIC_RAW' from 'time' (unknown location)


No test results were found.
Log directory:  Could not determine logfile location.
```

File: scripts/zfs-tests.sh (hunk 1 above)
Comment (szubersk) [reply scripts/zfs-tests.sh:711]: > ```shell
> { ${T_R}
>     -...
>     2>&1; RUNRESULT=$?; } | tee "$R_F"
> ```
The snippet above still spawns a subshell for the pipeline.

Comment (rincebrain) [reply scripts/zfs-tests.sh:711]: I'd be more than happy to hear less gross alternatives than stashing the return value in a file and retrieving it, I just don't currently see any.

Comment (nabijaczleweli) [reply scripts/zfs-tests.sh:711]: Yeah, the lint's right, I overlooked that. I don't think there's a better way than doing it this way with /bin/sh.

Comment (szubersk) [reply scripts/zfs-tests.sh:711]: I think using tempfile is fine. I don't see any other obvious way.
Lesson learned from this issue: we shouldn't overlook having `set -e` at the beginning of every shell script :)

File: scripts/zfs-tests.sh [hunk 2]
```diff
@@ -695,17 +695,21 @@ fi
 #
 # Run all the tests as specified.
 #
+EVILRES=$(mktemp -u -t test-runner-error)
```
Comment (nabijaczleweli) [inline scripts/zfs-tests.sh:698]: This is both (notionally) unsafe and doesn't work with coreutils mktemp:
```
mktemp: too few X's in template ‘test-runner-error’
```

Suggest
```sh
EVILRES=$(mktemp -t test-runner-error.XXXXXX)
```
(which does create /tmp/t-r-e.XXXXXX.abcdef on FreeBSD (but who cares) and Just Works on coreutils)

File: scripts/zfs-tests.sh (hunk 1 above)
Comment (nabijaczleweli) [inline scripts/zfs-tests.sh:711]: `read -r RUNRESULT < "$EVILRES"` saves you a subshell and you also want to `rm "$EVILRES"`

**Comment by freqlabs** [comment]
Maybe test-runner.py itself should have a log file option instead of this tee hack.

**Comment by freqlabs** [comment]
Even better though would be reworking the test runner to use the python logging library, which should be able to handle this and the other quirky output kludges littered throughout the test runner.

For now I'm fine with working around it in the shell script, I don't want to hijack your PR, but it would sure be nice to clean up the test runner one of these days.

**Comment by szubersk** [comment]
I think it would be beneficial to deliver a hot fix in this PR ASAP (since FreeBSD checks are literally broken right now) and focus on a long-term solution in a different PR.
The workaround in here looks OK to me, after applying nab comments we should be good to go.

**Comment by szubersk** [comment]
> Even better though would be reworking the test runner to use the python logging library, which should be able to handle this and the other quirky output kludges littered throughout the test runner.

`logging` offers Handlers mechanism to implement tee-ing https://docs.python.org/3/howto/logging.html#logging-advanced-tutorial

**Comment by behlendorf** [comment]
> I'd probably drop it unless someone chimes in with a justification, personally, I just didn't have any desire to start a debate over it versus fixing the immediate problem here.

I'd suggest dropping it as well.

**Comment by freqlabs** [comment]
To be clear, all we really need to do in order to fix the tests on FreeBSD right now is to s/_RAW// in test-runner.py.in and the rest of the changes are not an immediate concern.

**Comment by behlendorf** [comment]
If we can get a PR open with the [minimal fix](https://github.com/openzfs/zfs/pull/12990#issuecomment-1018888856) we should be able to get it merged without much delay.

**Comment by nabijaczleweli** [comment]
@behlendorf #12995

**Comment by rincebrain** [comment]
I wasn't really expecting "stop the test runner from silently succeeding" to be controversial or I would have just submitted the even more trivial fix.

**Comment by behlendorf** [comment]
We should definitely fix that too!  Let me just go ahead and merge the tiny fix to get the FreeBSD bots back on their feet ASAP.

**Comment by szubersk** [comment]
> I wasn't really expecting "stop the test runner from silently succeeding" to be controversial or I would have just submitted the even more trivial fix.

@rincebrain I don't think the change in here is controversial, it is actually a very good one. We still need the correction for `scripts/zfs-tests.sh` you made here or @freqlabs change in `test-runner.py` (#12995 just fights fire with a logic change). The issue with `zfs-tests.sh` ignoring `test-runner.py` exit code hasn't been not tackled yet at all. Abandoning changes here seems like a waste.

It's not a bad thing that people pitch in ideas. Sometimes the discussion becomes chaotic, unfortunately.

**Comment by rincebrain** [comment]
It is a waste.

@szubersk I've reread your reply a few times, and I can't really interpret it, especially giving an explanation of a one-line fix to someone in the PR it originated in as anything but extreme condescension. Please don't talk down to people, it's never constructive.

**Pull request closed without merging**